asf_heat_pump_suitability/pipeline/preweight_error_analysis/preweight_tenure_error_analysis.py

```diff
@@ -0,0 +1,457 @@
+        {"TENURE": "tenure", "UPRN": "id"}
+    )
+    return sample
+
+
+def load_target(target_path):
+    """
+    Load the target dataset from the census data.
+
+    Args:
+        target_path (str): The path to the Excel file containing the census data.
+
+    Returns:
+        tenure (pl.DataFrame): The preprocessed census data, with irrelevant rows and columns removed, and columns renamed.
+    """
+    fs = s3fs.S3FileSystem()
+    with fs.open(target_path, mode="rb") as f:
+        content = f.read()
+    tenure = pl.read_excel(content, sheet_name="3c", engine="calamine")
+    tenure = tenure.rename(tenure[2].to_dicts().pop()).slice(3)
+    tenure = tenure.rename(
+        {
+            "Area Code": "lsoa",
+            "Owned or shared ownership": "owner-occupied",
+            "Social Rented": "rental (social)",
+            "Private Rented or lives rent free": "rental (private)",
+        }
+    )
+    censored_vals = 0
+    int_cols = ["owner-occupied", "rental (private)", "rental (social)"]
+    tenure = tenure.with_columns(
+        [pl.col(int_cols).str.replace("c", f"{censored_vals}").cast(pl.Int64)]
+    ).select(pl.col(["lsoa", "owner-occupied", "rental (private)", "rental (social)"]))
+    return tenure
+
+
+def calculate_proportions(counts: dict) -> dict:
+    """
+    Calculate proportions for each key in a dictionary.
+    Args:
+        counts (dict): dictionary with counts for each key
+    Returns:
+        proportions (dict): dictionary with proportions for each key
+    """
+    total = sum(counts.values())
+    proportions = {k: v / total for k, v in counts.items()}
+    return proportions
+
+
+def calculate_diffs(dict1: dict, dict2: dict) -> dict:
+    """
+    Calculate differences between two dictionaries.
+    Args:
+        dict1 (dict): first dictionary
+        dict2 (dict): second dictionary
+    Returns:
+        differences (dict): dictionary with differences for each key
+    """
+    keys = set(dict1) | set(dict2)
+    differences = {k: dict1.get(k, 0) - dict2.get(k, 0) for k in keys}
+    return differences
+
+
+def calculate_rmse(proportions_diff: dict) -> float:
+    """
+    Calculate Root Mean Square Error (RMSE) from a dictionary of differences.
+    Args:
+        proportions_diff (dict): dictionary with differences for each key
+    Returns:
+        float: RMSE
+    """
+    squares = [x**2 for x in proportions_diff.values()]
+    mean_square = np.mean(squares)
+    rmse = np.sqrt(mean_square)
+    return rmse
+
+
+def calculate_mae(proportions_diff: dict) -> float:
+    """
+    Calculate Mean Absolute Error (MAE) from a dictionary of differences.
+    Args:
+        proportions_diff (dict): dictionary with differences for each key
+    Returns:
+        float: MAE
+    """
+    absolute_values = [abs(x) for x in proportions_diff.values()]
+    mae = np.mean(absolute_values)
+    return mae
+
+
+def process_single_lsoa(
+    sample: pl.DataFrame, tenure: pl.DataFrame, test_lsoa: str
```

Comment:
this is so close to being usable across features (i.e. not just tenure) so perhaps worth making it so?

i.e. remove mention of 'tenure' from the variable names and put 'target' instead, and add an argument for the feature column, e.g. `feature_name: str =TENURE`